Add email campaign name [MAILPOET-5646]

costasovo opened the pull request:
## Description

This PR adds campaign name input into the new email editor, and it also adds support for the campaign name in other parts of the plugin, where we refer to an email by name.

#### Campaign Name in the editor

![Screenshot 2023-12-13 at 15 47 02](https://github.com/mailpoet/mailpoet/assets/1082140/13fcc88a-9c45-409e-96cc-118e0b9e0c47)


#### Campaign in other parts of the plugin

1. In emails listing, when an email has a campaign name we show the campaign name and a subject is displayed on the other line in smaller font.
<img width="726" alt="Screenshot 2023-11-13 at 10 05 04" src="https://github.com/mailpoet/mailpoet/assets/1082140/9d097ef3-6520-471f-992c-adbeab07642f">

2. List of emails in dynamic segments UI. For segments where a user needs to pick segment we display only the campaign name if available.
3. The page title in email stats page – we display both name and subject.

## Code review notes

@JanJakes I chose to store the campaign name as a title in `wp_post` table. I think this makes sense for the future reusability of the editor. In order to support the campaign name in listings and other places in the plugin where we use Doctrine I added a minimalistic Doctrine entity for the WPPost. To support the entity I needed to do a few changes in the metadata factory for table prefixes. I put some logic directly to the factory class which makes this class aware that some tables are from WordPress. We may want to add an abstraction, but as this is a project (not a library) I thought it is ok.  I want to ask you for review of the related commits: c5c4e47f5094e8a02fcbb5b1585425a737c33844, fc2103661e2c83d4de7ab9ef9594629987c3ccf5 

## QA notes

1. Go to the experimental page and enable the new email editor (`wp-admin/admin.php?page=mailpoet-experimental`)
2. Create a standard newsletter in the new editor and set the campaign name and title
3. Verify that the campaign name is displayed in all places as expected
4. Are there some places where we should email name I missed?

## Linked PRs

https://github.com/mailpoet/mailpoet-premium/pull/820

## Linked tickets

[MAILPOET-5646]

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-5646]: https://mailpoet.atlassian.net/browse/MAILPOET-5646?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

costasovo committed:
Improve the name of the action for updating mailpoet-specific data

The original name suggested that the method could be used for
updating any property of a full email post entity but actually updates only mailpoet
specific data.

[MAILPOET-5646]

costasovo committed:
Allow title for the email custom post

[MAILPOET-5646]

costasovo committed:
Add a document actions panel for email name editing

[MAILPOET-5646]

costasovo committed:
Show the campaign name alongside the subject in standard emails listing

[MAILPOET-5646]

costasovo committed:
Improve table prefixing to allow Doctrine entities for WP tables

WordPress tables use only the DB prefix, but MailPoet tables use
DB prefix + plugin prefix (mailpoet).

This commit changes TablePrefixMetadataFactory to be able to distiguish
WP table for posts and adds proper prefix.
[MAILPOET-5646]

costasovo committed:
Add minimalistic WpPostEntity for fetching basic post data for listing

[MAILPOET-5646]

costasovo committed:
Fetch campaign titles from wp_posts in newsletters listing

[MAILPOET-5646]

costasovo committed:
Allow sorting by name + subject in newsletters listing

[MAILPOET-5646]

costasovo committed:
Use campaign name if available in newsletters listing in segment editor

[MAILPOET-5646]

costasovo committed:
Display campaign name and subject on the newsletter stats page

[MAILPOET-5646]

costasovo committed:
Extend the editor acceptance test by checking campaign name functionality

[MAILPOET-5646]

JanJakes commented:
> @JanJakes I chose to store the campaign name as a title in wp_post table. I think this makes sense for the future reusability of the editor. In order to support the campaign name in listings and other places in the plugin where we use Doctrine I added a minimalistic Doctrine entity for the WPPost. To support the entity I needed to do a few changes in the metadata factory for table prefixes. I put some logic directly to the factory class which makes this class aware that some tables are from WordPress. We may want to add an abstraction, but as this is a project (not a library) I thought it is ok.

@costasovo Interesting solution, I've never thought of that (adding entities for some or all WP tables), but it seems that so far, we haven't really had a strong need. Now, I wonder — should we start using them in our codebase? And if so, in which cases yes, and which cases no?

Maybe it comes down to why do you need the entity in this case, and, more broadly, what's the plan for storing newsletters in the future. Currently, old emails have the `NewsletterEntity`, while emails for the new editor are saved in WP posts, I suppose. But why do we need to join them? Do you also create a `NewsletterEntity` record for each email that is stored in WP posts?

Without a deeper understanding, I tend to think it would be better to avoid having a Doctrine entity for WP posts, so we don't use it in more and more places, and rather stick to WP-native functionality. For instance, to fetch multiple post titles, could we just use `get_posts` or `WP_Query` with some arguments?

> I want to ask you for review of the related commits: https://github.com/mailpoet/mailpoet/commit/c5c4e47f5094e8a02fcbb5b1585425a737c33844, https://github.com/mailpoet/mailpoet/commit/fc2103661e2c83d4de7ab9ef9594629987c3ccf5

Should I only check the WP post entity part, or do the full PR review?

costasovo commented:
@JanJakes Thank you for all these questions. They helped to challenge this idea. Maybe there could be a better approach...

> @costasovo Interesting solution, I've never thought of that (adding entities for some or all WP tables), but it seems that so far, we haven't really had a strong need. 

The need comes from the idea of storing the campaign name as the title of the post in combination with the fact that we want to present the campaign name as the name of an email (replacing the subject) in a couple of places in the plugin. Now why do I want to store the campaign name as a title? 
The editor will be used in more projects, and I want to provide a way to set the title of the email document, and saving in the post's title column feels like a natural way of doing this. In the editor, there is no problem with this. The main issue comes from the listing repository, where we build on top of the doctrine's DQL, and we need to cover sorting and searching. Without an entity, I can't refer to/fetch the value using DQL.

> Now, I wonder — should we start using them in our codebase? And if so, in which cases yes, and which cases no?

I would avoid using them as much as possible and limit the usage only to reading the data. Usage for updates would cause skipping various WP hooks. Could mark the entity as [Read Only be sufficient](https://www.doctrine-project.org/projects/doctrine-orm/en/2.16/reference/improving-performance.html#read-only-entities)?

> Maybe it comes down to why do you need the entity in this case, and, more broadly, what's the plan for storing newsletters in the future. Currently, old emails have the NewsletterEntity, while emails for the new editor are saved in WP posts, I suppose. 

I'm trying to distinguish between common email data vs MailPoet specific email data. I would like to store the common data (content, title, subject?, preheader?) in the wp posts infrastructure so that the editor can be used without MailPoet tables but still want to use MailPoet tables for MailPoet-specific stuff and also for email processing logic (sending queue, scheduled tasks, etc.). 

> But why do we need to join them? Do you also create a NewsletterEntity record for each email that is stored in WP posts?

I'm joining them (in the sense of linking them with a foreign key) because of the sending processing as it works with the NewsletterEtity, yes I create a NewsletterEntity for every post that has mailpoet_email type. As for joining on the query level, I used that in the listing query and one other query used for fetching lists of emails for the dynamic segment builder.

> Without a deeper understanding, I tend to think it would be better to avoid having a Doctrine entity for WP posts, so we don't use it in more and more places, and rather stick to WP-native functionality. For instance, to fetch multiple post titles, could we just use get_posts or WP_Query with some arguments?

I was trying to hold to the approach of relying only on WP-native functionality, but then I ran into the issue with the listing repository. So I was thinking we need an exception for this case. I agree that we should rather avoid doing this. This makes me think that we could make an exception the other way around. I mean, we could add a column for the campaign name to the newsletters table and sync the title into it (e.g. in a post's update hook). This would also help with fetching the names for lists and might be less "dangerous" than introducing an entity for a WP post. @JanJakes , @lysyjan What do you think about the alternative?

lysyjan commented:
> I would avoid using them as much as possible and limit the usage only to reading the data. Usage for updates would cause skipping various WP hooks. Could mark the entity as [Read Only be sufficient](https://www.doctrine-project.org/projects/doctrine-orm/en/2.16/reference/improving-performance.html#read-only-entities)?

This approach sounds good to me, but if we want to avoid having entities for WP posts, we can choose the approach with a copied campaign name in our newsletter table. I am not a fan of duplicated data, but I can understand that if needed.

JanJakes commented:
@costasovo Thanks for the details.

> still want to use MailPoet tables for MailPoet-specific stuff and also for email processing logic (sending queue, scheduled tasks, etc.).

I understand we'll need that now, so we don't have to rewrite all the existing logic, but in the final state, shouldn't the `NewsletterEntity` no longer exist? I don't know what data you consider to be MailPoet-specific, but WP posts allow you to save pretty much anything as post metadata, and things like tasks and queues are not stored in the `NewsletterEntity` (in other words, they can reference the WP post ID directly.)

In any case, in the meantime, I think it's fine to use the entity and mark it readonly, if that makes the code significantly simpler. Rather than a query hint, we can [annotate that way the entity itself](https://www.doctrine-project.org/projects/doctrine-orm/en/2.16/reference/attributes-reference.html#entity). I still don't fully understand why it's really required, as using `get_posts` or `WP_Query` for the titles seems to be simple as well (it's not a big deal to have two queries for the listing data, I suppose), but ultimately, I'm fine with any of these solutions.

Last food for thought — WP posts can be extended by 3rd party plugins for fine-grained access control, etc. If we want to preserve that functionality, we'd need to stick to `get_posts` or `WP_Query` to avoid bypassing 3rd party hooks. In that case, you can do one WP query to fetch all IDs, titles, timestamps, and such, and then JOIN our entity on that.

costasovo committed:
Mark WpPostEntity as read-only

Read-only was added to prevent using the entity for modifying post
data.
We want to allow modifying posts only via WP functions so that we don't skip any hooks etc.
[MAILPOET-5646]

costasovo commented:
@lysyjan I discussed the issue with @JanJakes, and we will go with the Read-only entity to avoid duplicate data.

lysyjan commented on mailpoet/assets/js/src/email-editor/engine/components/header/document-actions.tsx:
```diff
@@ -0,0 +1,86 @@
+
+// @see https://github.com/WordPress/gutenberg/blob/5e0ffdbc36cb2e967dfa6a6b812a10a2e56a598f/packages/edit-post/src/components/header/document-actions/index.js
+
+export function DocumentActions() {
```
The name of this component is a bit confusing to me. Why did you choose it? I'm expecting it's based on the Gutenberg component but I think we could use something more related to the functionality of the component.

lysyjan commented on mailpoet/assets/js/src/email-editor/engine/components/header/index.scss:
```diff
@@ -0,0 +1,40 @@
+// Document actions - Popup for editing email/campaign title
+.mailpoet-email-editor-document-actions__dropdown {
+  .components-popover__content {
+    min-width: 240px;
```
It's only a detail, but I have checked the Figma design, and it contains a higher width. Do you have some reason for using 240px instead of 280px?

costasovo replied:
I copied the style from automations. 280px looks nicer. I'll update it.

lysyjan commented on mailpoet/assets/js/src/email-editor/engine/components/header/index.scss:
```diff
@@ -0,0 +1,40 @@
+    padding: 16px;
+  }
+
+  .mailpoet-email-editor-dropdown-name-edit-title {
```
From what I could see in the Figma, the title should be upper case and bold. Nevertheless, I checked the Automation, and there is a similar component, so I am not sure about that. In my opinion, at least a bold font would be fine here. What do you think?

costasovo replied:
Nice catch! It doesn't even have to be separately. Moving the text into the TextControl label will make it look as in the Figma design.

lysyjan commented on mailpoet/lib/AdminPages/Pages/DynamicSegments.php:
```diff
@@ -186,6 +186,7 @@ private function getNewslettersList(): array {
       $result[] = [
         'id' => (string)$newsletter->getId(),
         'subject' => $newsletter->getSubject(),
+        'name' => $newsletter->getPresentableName(),
```
NewsletterResponseBuilder uses `campaign_name` and here is `name`. Is there a reason for that? I think we should unify the used name for all places. It can be confusing to mix those two.

lysyjan reviewed (commented):
I'm sorry for the late review. Both of us have different duties in the last weeks.
I know how complicated it was to solve the problem with WP posts.
I have only a few minor remarks, I guess.

costasovo committed:
Rename the component for setting the campaign name

Let's go with the specific name. We can come up with a more abstract name
as we add more settings to the component.
[MAILPOET-5646]

costasovo committed:
Increase width of the campaign name dropdown

[MAILPOET-5646]

costasovo committed:
Move campaign name title to TextControl label

[MAILPOET-5646]

costasovo committed:
Improve name of a getter for campaign name or subject on NewsletterEntity

[MAILPOET-5646]

costasovo commented:
@lysyjan I've rebased the branch and addressed all the comments in the new commits.

lysyjan reviewed (dismissed):
👍

Aschepikov commented:
Hi, @costasovo I found some issues with the gutenberg editor, some of them are on the prod maybe after the refactoring, Pls check it and tell me what can be fixed here and for what we need a separate ticket.

1. The sender is always empty,  there is no auto-filling like in the current newsletter editor. It is present on the prod too.
![image](https://github.com/mailpoet/mailpoet/assets/113185437/81962dd7-2f67-49f3-a524-72051d81f314)

2. We are implementing the campaign name and the logic is that there is a campaign name and underneath there is a subject text. If I am creating a campaign name on the email editor step
![image](https://github.com/mailpoet/mailpoet/assets/113185437/fc9311a5-21b2-4096-836f-9f259f8fc62d)
So in the next step, I will see the send page 
![image](https://github.com/mailpoet/mailpoet/assets/113185437/b1f5a647-a6e4-44bb-a20d-aedcad08b17f)
Without a campaign name but with the subject and preview text
![image](https://github.com/mailpoet/mailpoet/assets/113185437/aceb1ce0-b958-491f-823b-4a9155a7db6b)
With the filled campaign name, subject, and preview I will see this name in the list with the default campaign name (the new one wasn't saved) and the subject so we need to add the campaign name on the send page with the subject and without preview text.
![image](https://github.com/mailpoet/mailpoet/assets/113185437/7e39d19c-242f-463c-8b2d-6a3b7345195b)

So the campaign name is not saving it looks like there is no autosaving at all, but last time I checked it was ok. And the campaign name should be on the send page too.

3. The sending of the Gutenberg emails isn't working, I am checking the last versions to understand where it was implemented, the last time sending was ok
![image](https://github.com/mailpoet/mailpoet/assets/113185437/c5a775c8-c350-462e-b520-a70b0e19ddd3)
![image](https://github.com/mailpoet/mailpoet/assets/113185437/e7741100-57e4-4006-aa6c-4191e3a1a4c9)
Can it be connected that the email wasn't saved and is empty?

Aschepikov commented:
For everything else it is possible to check when the campaign name will be saved and the email will be sent

costasovo commented:
@Aschepikov Thank you for the thorough review 👍 
ad 1) This seems like a bug. Probably not related to the PR but I'll look into it.
ad 2) We will get rid of the send page in later phases of the new email editor, but I see it could be confusing. I suggest we add the campaign name to the send page as a no-editable heading. That way it would be present in all steps. What do you think? @pavel-mailpoet Any other ideas?
ad 3) I'll look into it.

pavel-mailpoet commented:
> ad 2) We will get rid of the send page in later phases of the new email editor, but I see it could be confusing. I suggest we add the campaign name to the send page as a no-editable heading. That way it would be present in all steps. What do you think? @pavel-mailpoet Any other ideas?

Hmm, we will remove the send page. However, we want to release the editor as a beta with the send page. So some users will still using it for a bit and so we should make it usable. The solution you suggested sounds easy enough to do  and could improve this feature for the users.

costasovo committed:
Refactor construction of NewsletterEntity when creating an email in the new editor

When we create a new email in the new editor we ensure there is
also an associated NewsletterEntity.
This commit improves this functionality by replacing custom code with
 NewsletterSaveController, which ensures that the sender name and address are set.
[MAILPOET-5646]

costasovo committed:
Add campaign name to the sending page

[MAILPOET-5646]

costasovo committed:
Refactor Send button to an extra component

[MAILPOET-5646]

costasovo committed:
Improve isEmpty selector to check also campaign name

[MAILPOET-5646]

costasovo committed:
Disable send button when saved content is empty

[MAILPOET-5646]

costasovo commented:
@lysyjan @Aschepikov I added five commits addressing the issues.
ad 1) Fixed but only for newly created emails. We are still behind a feature flag so I doesn't make sense adding a migration.
ad 2) I added a heading at the top of the send page. The heading is displayed only if the email has a campaign name.
ad 3) The problem with sending is related to the empty email body. I added a check that send button is disabled in case the **saved body** is empty. Please note this is probably a temporary fix because we plan to add autosave and also email content validation when user clicks on the send button.

Aschepikov commented:
HI @costasovo thank you for the fixes. Now when I am duplicating the email there are creating two copies of it, one with the copy of campaign name and one with the copy of campaign name and subject 
<img width="1672" alt="image" src="https://github.com/mailpoet/mailpoet/assets/113185437/c64174cc-5b34-41ca-8c6b-2fcf9b653146">

pavel-mailpoet commented:
@Aschepikov I don't think this problem is related to this branch. We have a ticket for it already: https://mailpoet.atlassian.net/browse/MAILPOET-5810

Aschepikov reviewed (approved):
GJ!
